packages/flutter/lib/src/widgets/animated_scroll_view.dart

```diff
@@ -546,51 +675,136 @@ abstract class _AnimatedScrollViewState<T extends _AnimatedScrollView> extends S
-    _sliverAnimatedMultiBoxKey.currentState!.insertAllItems(index, length, duration: duration);
+    if (widget.removedSeparatorBuilder == null) {
+      _sliverAnimatedMultiBoxKey.currentState!.insertAllItems(index, length, duration: duration);
+    } else {
+      final int itemIndex = _computeItemIndex(index);
+      final int lengthWithSeparators = _itemsCount == 0 ? length * 2 - 1 : length * 2;
+      _sliverAnimatedMultiBoxKey.currentState!.insertAllItems(itemIndex, lengthWithSeparators, duration: duration);
+    }
   }
 
-  /// Remove the item at `index` and start an animation that will be passed to
-  /// `builder` when the item is visible.
+  /// Remove the item at [index] and start an animation that will be passed to
+  /// [builder] when the item is visible.
+  ///
+  /// If using [AnimatedList.separated], the animation will also be passed to the
+  /// corresponding separator's [AnimatedList.removedSeparatorBuilder].
   ///
   /// Items are removed immediately. After an item has been removed, its index
-  /// will no longer be passed to the `itemBuilder`. However, the
-  /// item will still appear for `duration` and during that time
-  /// `builder` must construct its widget as needed.
+  /// will no longer be passed to the [builder]. However, the
+  /// item will still appear for [duration] and during that time
+  /// [builder] must construct its widget as needed.
   ///
   /// This method's semantics are the same as Dart's [List.remove] method: it
   /// decreases the length of items by one and shifts all items at or before
-  /// `index` towards the beginning of the list of items.
+  /// [index] towards the beginning of the list of items.
   ///
   /// See also:
   ///
   ///   * [AnimatedRemovedItemBuilder], which describes the arguments to the
-  ///     `builder` argument.
+  ///     [builder] argument.
   void removeItem(int index, AnimatedRemovedItemBuilder builder, { Duration duration = _kDuration }) {
-    _sliverAnimatedMultiBoxKey.currentState!.removeItem(index, builder, duration: duration);
+    final AnimatedRemovedSeparatorBuilder? removedSeparatorBuilder = widget.removedSeparatorBuilder;
+    if (removedSeparatorBuilder == null) {
+      _sliverAnimatedMultiBoxKey.currentState!.removeItem(index, builder, duration: duration);
+    } else {
+      final int itemIndex = _computeItemIndex(index);
+      _sliverAnimatedMultiBoxKey.currentState!.removeItem(itemIndex, builder, duration: duration);
+      if (_itemsCount > 1) {
+        if (itemIndex == _itemsCount - 1) {
+          // Removing from the end of the list, so the separator to remove is the one at `last index` - 1.
+          _sliverAnimatedMultiBoxKey.currentState!.removeItem(itemIndex - 1, _toRemovedItemBuilder(removedSeparatorBuilder, index - 1), duration: duration);
+        } else {
+          _sliverAnimatedMultiBoxKey.currentState!.removeItem(itemIndex, _toRemovedItemBuilder(removedSeparatorBuilder, index), duration: duration);
```

Comment:
Adding comments for each of these calls to removeItem would be helpful, like 'this removes the item', 'this removes the separator', etc. :)

Reply:
I added comments. Do they help, or do we need to improve them further? If so, please let me know what could help make this easier to understand.

Reply:
They help! Thank you!